### PR TITLE
Add remove method to PMT

### PR DESCRIPTION
**Pull request by diegomrsantos**

### CODE REVIEW

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTAcc.java [hunk 1]
```diff
@@ -72,6 +72,7 @@
 
   private List<PMTNode> visitedAcc = new ArrayList<>();
```
Comment (michalzajda) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTAcc.java:73]: this should be renamed to outdated or similar and trigger removal after add

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java [hunk 2]
```diff
@@ -122,13 +147,116 @@ public void insertNode(
...
+      acc.setTip(null);
+    } else {
+      acc.remove(this);
+      var newLeaf = new PMTLeaf(new PMTKey(new byte[0]), this.getValue());
+      acc.add(newLeaf);
+      acc.setTip(newLeaf);
+    }
+  }
+
+  private void handleBranchWithTwoChildren(PMTAcc acc, Function<byte[], PMTNode> read, PMTKey key) {
```
Comment (michalzajda) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java:212]: this is only for no value branch. Maybe it should be somehow merged with branchDoesNotHaveValue check?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java [hunk 3]
```diff
@@ -122,13 +147,116 @@ public void insertNode(
...
-          String.format("Unexpected subtree: %s", commonPath.whichRemainderIsLeft()));
+          String.format(UNEXPECTED_SUBTREE_ERROR_MSG, commonPath.whichRemainderIsLeft()));
+    }
+  }
+
+  @Override
+  public void removeNode(
+      PMTKey key, PMTAcc acc, Function<PMTNode, byte[]> represent, Function<byte[], PMTNode> read) {
+    final PMTPath commonPath = PMTPath.findCommonPath(this.getKey(), key);
+    if (commonPath.whichRemainderIsLeft() == PMTPath.RemainingSubtree.NONE) {
```
Comment (michalzajda) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java:158]: to be exact this is caused by empty key. And the NONE remaining remainders is the consequence. Maybe lets have explicit condition?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMT.java [hunk 4]
```diff
@@ -237,6 +237,28 @@ public PMT add(byte[] key, byte[] val) {
...
+      this.root.removeNode(pmtKey, acc, this::represent, this::read);
+      if (acc.notFound()) {
+        return this;
+      } else {
+        PMTNode newRoot = acc.getTip();
+        for (PMTNode sanitizedAcc : acc.getNewNodes().stream().filter(Objects::nonNull).toList()) {
+          byte[] serialisedNode = this.pmtNodeSerializer.serialize(sanitizedAcc);
+          if (hasDbRepresentation(serialisedNode)) {
+            this.db.save(hash(serialisedNode), serialisedNode);
+          }
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMT.java:253]: This task is only to add remove API to the tree

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTAcc.java (hunk 1 above)
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTAcc.java:73]: I believe this isn't related to the current task

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java (hunk 2 above)
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java:212]: `branchDoesNotHaveValue` is used in another case as well.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java [hunk 5]
```diff
@@ -122,13 +147,116 @@ public void insertNode(
...
+      acc.remove(this);
+      var branchWithNext = new PMTBranch(this);
+      branchWithNext.setEmptyChild(key.getFirstNibble());
+      acc.add(branchWithNext);
+      acc.setTip(branchWithNext);
+    }
+  }
+
+  private void handleBranchWithOnlyOneChild(PMTAcc acc) {
+    if (branchDoesNotHaveValue()) {
```
[the remark this answers is not in the document]
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java:201]: I think these are two different things.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java (hunk 3 above)
Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java:158]: Not sure if I get it, this is the same logic used in `insertNode`.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java (hunk 2 above)
Comment (michalzajda) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java:212]: well, in one case it's outside, in the other inside.

Consider combining value + children computation for simplicity, as mentioned in comment above. These are different, but identify need for collapsing the branch well

Comment (diegomrsantos) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java:212]: I'll rename `handleBranchWithTwoChildren` to `handleBranchWithTwoChildrenAndNoValue`

**Pull request merged**